PLOT: Split matplotlib specific code from pandas plotting

datapythonista opened the pull request:
- [X] xref #14130
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

In order to be able to change pandas plotting backend, I think we need 3 main things:
1. Decouple the current matplotlib backend from the pandas plotting system, and just call it using a defined public API
2. Implement an option to select the backend (e.g. `pandas.set_option('plotting.backend', 'pandas.matplotlib')`)
3. Have other alternative backends which define the same API as the current matplotlib one

This PR is splitting most of the matplotlib backend code into a new `_matplotlib.py` file. Some matplotlib code still lives in `plotting/_core.py`, but would be nice to have early feedback on whether everybody is happy with this approach before moving forward. CC: @jreback @TomAugspurger

Commit:
First split of matplotlib functions from pandas plotting core

TomAugspurger commented:
Thanks, will take a look later. In case it's useful now, I started https://github.com/TomAugspurger/pandas/commit/b07aba28a37b0291fd96a1f571848a7be2b6de8d, which adds a config option for plotting, but never got to submitting a PR.

jbrockmendel commented:
Totally on-board with the concept.  What other backends do you expect (either implemented here or downstream)?  My understanding is many of the other graphics libraries wrap matplotlib themselves.

TomAugspurger commented:
I know of two libraries that have reimplemented pd.DataFrame.plot

- https://github.com/altair-viz/pdvega/
- https://hvplot.pyviz.org

datapythonista commented:
I think Phillip started hvplot with the discussions of this issue, but as we didn't finish this work, it ended monkeypatching pandas.

The goal here is that projects like the ones Tom mentioned, and others like https://github.com/PatrikHlobil/Pandas-Bokeh can be implemented simpler and in a standard way, so we make the life of devs and users easier.

Personally, I'd at some point split the matplotlib backend from pandas and have it as a third party package. But that's for the long term, and unrelated to this PR.

Commit:
Splitting matplotlib backend code from pandas plotting

Commit:
isort

datapythonista commented:
For what I could see in the code, the API that pandas plotting backends need to provide are:

- `LinePlot`
- `BarPlot`
- `BarhPlot`
- `HistPlot`
- `BoxPlot`
- `KdePlot`
- `AreaPlot`
- `PiePlot` (only for `pandas.Series`)
- `ScatterPlot` (only for `pandas.DataFrame`)
- `HexBinPlot` (only for `pandas.DataFrame`)
- `hist_series`
- `hist_frame`
- `boxplot_frame`
- `boxplot_frame_groupby`

The implementation in this PR moves all the `matplotlib` code into `pandas.plotting.matplotlib`. And keeps all the public documentation in pandas (outside of the matplotlib backend). In a follow up PR should be simple to add an option to change the backend to another respecting the same API as the matplotlib one.

Any comment appreciated.

CC: @tacaswell, @jakevdp, @philippjfr, @PatrikHlobil

Commit:
Removing boxplot import from __init__.py

jakevdp commented:
This is great - thanks for working on it!

Commit:
Import plot backend in a lazy way

TomAugspurger commented on pandas/plotting/matplotlib/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from .boxplot import BoxPlot, boxplot_frame, boxplot_frame_groupby
```
Rename this module to `pandas/plotting/_matplotlib`

TomAugspurger commented on pandas/plotting/matplotlib/main.py:
```diff
@@ -0,0 +1,1374 @@
+import re
```
Maybe "core" instead of "main"? I typically associate main with `__main__` / running from the command line.

datapythonista committed:
Merge remote-tracking branch 'upstream/master' into plot

datapythonista committed:
Renaming files as per review, and isort

tacaswell commented:
I am too :zzz: right now to have much sensible comments on the code.

If you are interested in pulling this whole thing out and putting it in `matplotlib/matplotlib-pandas` I am  open to that and would happily make sure anyone from the pandas side who needs / wants access will have it.  That is probably something to look at down the road after sorting out what you want the interface to be (mostly to avoid obligatory coordinated releases).

If there is interest in more aggressive re-thinking of the API https://github.com/DistrictDataLabs/yellowbrick/issues/826 , https://github.com/matplotlib/matplotlib/pull/14058, and https://paper.dropbox.com/doc/Matplotlib-4.0--AdQMb0ugRPWeqq3U4dEAHuSDAQ-WTYwd0NQaSHTjtLUZwkNx may be of interest.

Commit:
Merge remote-tracking branch 'upstream/master' into plot

Commit:
Splitting the rest of public plotting functions (the accessors were already moved) into the matplotlib specific and the pandas plotting stuff

pep8speaks commented:
Hello @datapythonista! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/plotting/__init__.py`](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/__init__.py):

> [Line 13:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/__init__.py#L13): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)









* In the file [`pandas/plotting/_matplotlib/core.py`](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_matplotlib/core.py):

> [Line 1007:50](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_matplotlib/core.py#L1007): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 1008:50](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_matplotlib/core.py#L1008): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent











* In the file [`pandas/plotting/_misc.py`](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py):

> [Line 40:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L40): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 128:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L128): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 129:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L129): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 130:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L130): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 196:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L196): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 236:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L236): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 343:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L343): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 344:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L344): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 345:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L345): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 346:80](https://github.com/pandas-dev/pandas/blob/5617b37fee52fee389b679a3c85d28d2fc462759/pandas/plotting/_misc.py#L346): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)

jreback reviewed (changes requested):
is this PR *just* moving things (and fixing the imports). would greatly appreciate that. Then can followon with a PR to change code. It is impossible to see what you are actually changing otherwise.

datapythonista commented:
100% agree, there is no functional change, and no public API change in this PR. Just moving all the code specific to matplotlib from `pandas/plotting/*` to `pandas/plotting/_matplotlib/*`. All the calls to the "public" functions in `_matplotlib` are done by importing the module in a lazy way (that fails when matplotlib is not installed in the same was as it fails now).

Ideally would like to avoid touching any tests to guarantee that no unintentional functional change happens. In practice I need to update the import paths of some tests, since they import private functions that have been moved.

Still work in progress, the change is much bigger than what I originally thought, but will provide more details on what the PR involves when it's ready to be merged.

jreback commented:
thanks @datapythonista yeah also happy to do this in pieces if its easier.

datapythonista committed:
Fixing imports and linting

datapythonista committed:
Fixing imports and restoring boxplot as a public method

datapythonista committed:
Fixing bug on warning stacklevel of moved function

datapythonista committed:
Avoiding imports inside pandas.plotting (pandas should only import public plotting objects)

datapythonista committed:
Merge remote-tracking branch 'upstream/master' into plot

datapythonista committed:
Working on fixing the duplicate plots problem

datapythonista committed:
Avoid importing _matplotlib/ from plotting/_core.py

datapythonista committed:
Bugfixing (only 2 tests failing)

datapythonista committed:
Fixing last failing plotting tests

datapythonista committed:
Merging from master

datapythonista committed:
Only importing matplotlib in tests when it's installed

datapythonista committed:
Not importing matplotlib in hist tests if it's not installed

datapythonista commented:
This should be ready now. This PR just moves code around, and shouldn't change the behavior of pandas at all. The implemented changes are:

- All the code that imports and uses matplotlib now lives in `pandas/plotting/_matplotlib`
- There is a clear API for what's in `pandas.plotting` and also in `pandas.plotting._matplotlib`, and no private methods are imported (except for tests, benchmarks and deprecated code)
- The changes in this PR should make very easy to change the matplotlib plotting backend by another one based in bokeh, holoviews, altair... as far as they are implemented using the same public API

Once this is merged I'll open issues to discuss the possible follow up changes:
- Using an option to specify the plotting backend
- Possible deprecations (there are up to 4 ways to plot the same thing in pandas, e.g. `df.plot(kind='box')`, `df.plot.box()`, `df.boxplot()` and `pandas.plotting.boxplot(df)`). Probably worth deprecating some, making the API for plotting backends simpler, and making users life easier by providing a single consistent way
- Possibly moving the matplotlib backend out of the pandas repo.
- Whether the backends should implement plots like `andrews_curves`, `autocorrelation_plot`,... or if they shouldn't be part of the pandas plotting API (but a third party).
- ...

jreback commented on pandas/plotting/_core.py:
```diff
@@ -2185,148 +324,111 @@ def plot_series(data, kind='line', ax=None,                    # Series unique
 
-def scatter_plot(data, x, y, by=None, ax=None, figsize=None, grid=False,
-                 **kwargs):
+def _get_plot_backend():
```
this should be the first function in this module and/or put somewhere else

jreback commented on pandas/plotting/_core.py:
```diff
@@ -2629,110 +594,159 @@ def boxplot_frame_groupby(grouped, subplots=True, column=None, fontsize=None,
+        grid=grid, ax=ax, figsize=figsize, layout=layout, sharex=sharex,
+        sharey=sharey, **kwds)
+
+
```
_get_plot_backend could be here (ignore my previous comment), to organize things. can you add an explanation block on what is the purpose of all this (also blank line separate before comments)

jreback commented on pandas/plotting/_matplotlib/__init__.py:
```diff
@@ -0,0 +1,16 @@
+from .boxplot import BoxPlot, boxplot, boxplot_frame, boxplot_frame_groupby
```
we tend to use absolute imports I think everywhere else

jreback reviewed (changes requested):
looks pretty good; on followup (or just create issues), pls move the tests as much as possible to a generic testing framework (for any backend), and mpl specific tests.

datapythonista committed:
Merge remote-tracking branch 'upstream/master' into plot

datapythonista committed:
Addressing PR comments (adding docstring, moving _get_plot_backend, and using absolute imports)

datapythonista commented:
Thanks for the review, addressed comments. Agree on splitting the tests, will follow up with that. I thought using relative imports made the matplotlib backend more decoupled (and easy to move if that's wanted at some point). Changed them to absolute imports for now, and we can see in the future if that's needed/useful.

datapythonista committed:
Merge remote-tracking branch 'upstream/master' into plot

jreback commented:
lgtm. @jorisvandenbossche or @TomAugspurger

TomAugspurger commented:
This seems to have lost our warning

```
In [1]: paste
import pandas as pd
import matplotlib.pyplot as plt
df = pd.Series(1, pd.date_range('2000', periods=12))

fig, ax = plt.subplots()

ax.plot(df.index, df.values)
## -- End pasted text --
/Users/taugspurger/sandbox/pandas/pandas/plotting/_converter.py:127: FutureWarning: Using an implicitly registered datetime converter for a matplotlib plotting method. The converter was registered by pandas on import. Future versions of pandas will require you to explicitly register matplotlib converters.

To register the converters:
        >>> from pandas.plotting import register_matplotlib_converters
        >>> register_matplotlib_converters()
  warnings.warn(msg, FutureWarning)
Out[1]: [<matplotlib.lines.Line2D at 0x10c614048>]
```

There's no warning on this branch.

I think that's because `import pandas` doesn't do an `import matplotlib` inside a try / except anymore. On master, `import pandas` would register our units / converters.

I don't recall when that deprecation is slated for removal. May as well be 0.24 though?

TomAugspurger commented:
On the converter warning, that was added in 0.21. https://github.com/pandas-dev/pandas/pull/24964. Something happened, and it wasn't actually showing up till 0.24. So I'm not sure what's best here.

If you do want to enforce the deprecation, then I'd recommend merging this first, and cleaning up the deprecation in a followup.

simonjayhawkins commented:
@datapythonista can you merge master.

Conflicting files
pandas/plotting/_compat.py

Commit:
Merging from master

Commit:
Merge branch 'plot' of https://github.com/datapythonista/pandas into plot

datapythonista committed:
Resolving error after merging master (invalid import)

datapythonista commented:
@TomAugspurger I restored the warning for now. The problem is that the `pandas.plotting._matplotlib` module is never loaded with your code, so registering the converters in that module, where it should happen is not possible.

I implemented it in a somehow hacky way. In `pandas.plotting` I test whether matplotlib is installed, and if it is, I register the converters (and load `pandas.plotting._matplotlib`). This is not idea, but we'll still be able to use other plotting backends without problems, and it's temporary (we can see later if we want to remove the automatic registration of converters in 0.25, 1.0 or when).

I think this solution is the best, let me know if you have a better idea (I'm not sure if this is what you were proposing).

jreback commented:
can you merge master

TomAugspurger commented:
>  I restored the warning for now.

Thanks. Looks good, and I'll do a followup issue to discuss removing it (I think after 0.25).

PatrikHlobil commented:
> This should be ready now. This PR just moves code around, and shouldn't change the behavior of pandas at all. The implemented changes are:
> 
> * All the code that imports and uses matplotlib now lives in `pandas/plotting/_matplotlib`
> * There is a clear API for what's in `pandas.plotting` and also in `pandas.plotting._matplotlib`, and no private methods are imported (except for tests, benchmarks and deprecated code)
> * The changes in this PR should make very easy to change the matplotlib plotting backend by another one based in bokeh, holoviews, altair... as far as they are implemented using the same public API
> 
> Once this is merged I'll open issues to discuss the possible follow up changes:
> 
> * Using an option to specify the plotting backend
> * Possible deprecations (there are up to 4 ways to plot the same thing in pandas, e.g. `df.plot(kind='box')`, `df.plot.box()`, `df.boxplot()` and `pandas.plotting.boxplot(df)`). Probably worth deprecating some, making the API for plotting backends simpler, and making users life easier by providing a single consistent way
> * Possibly moving the matplotlib backend out of the pandas repo.
> * Whether the backends should implement plots like `andrews_curves`, `autocorrelation_plot`,... or if they shouldn't be part of the pandas plotting API (but a third party).
> * ...

Hi, thanks for working on this issue. If there is native support for a plotting backend switch, implementing new plotting solutions for Pandas will be much more attractive! 

Does every backend need to provide a this APIs or are they optional? This could, in my opinion) discourage people to implement their backend solution + restrict the creativity of the developers because each backend could provide some innovative ideas (eg matplotlib vs seaborn or bokeh vs holoviews).

Furthermore, is there also a minimal set of keyword arguments (x, y, color, xlabel, ...) that has to be implemented for each glyph such that some minimal examples work for each backend? Also, will the development of the plotting backends happen inside pandas-dev, a new pandas-plotting-dev or a own repository like hvplot/Pandas-Bokeh?

Best, Patrik

TomAugspurger commented:
Backends should implement the same API, aside from the return type. For alternative APIs, libraries can accept DataFrames, just like seaborn.

> On Jun 2, 2019, at 21:31, Patrik Hlobil <notifications@github.com> wrote:
> 
> This should be ready now. This PR just moves code around, and shouldn't change the behavior of pandas at all. The implemented changes are:
> 
> All the code that imports and uses matplotlib now lives in pandas/plotting/_matplotlib
> There is a clear API for what's in pandas.plotting and also in pandas.plotting._matplotlib, and no private methods are imported (except for tests, benchmarks and deprecated code)
> The changes in this PR should make very easy to change the matplotlib plotting backend by another one based in bokeh, holoviews, altair... as far as they are implemented using the same public API
> Once this is merged I'll open issues to discuss the possible follow up changes:
> 
> Using an option to specify the plotting backend
> Possible deprecations (there are up to 4 ways to plot the same thing in pandas, e.g. df.plot(kind='box'), df.plot.box(), df.boxplot() and pandas.plotting.boxplot(df)). Probably worth deprecating some, making the API for plotting backends simpler, and making users life easier by providing a single consistent way
> Possibly moving the matplotlib backend out of the pandas repo.
> Whether the backends should implement plots like andrews_curves, autocorrelation_plot,... or if they shouldn't be part of the pandas plotting API (but a third party).
> ...
> Hi, thanks for working on this issue. If there is native support for a plotting backend switch, implementing new plotting solutions for Pandas will be much more attractive!
> 
> Does every backend need to provide a this APIs or are they optional? This could, in my opinion) discourage people to implement their backend solution + restrict the creativity of the developers because each backend could provide some innovative ideas (eg matplotlib vs seaborn or bokeh vs holoviews).
> 
> Furthermore, is there also a minimal set of keyword arguments (x, y, color, xlabel, ...) that has to be implemented for each glyph such that some minimal examples work for each backend? Also, will the development of the plotting backends happen inside pandas-dev, a new pandas-plotting-dev or a own repository like hvplot/Pandas-Bokeh?
> 
> Best, Patrik
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

TomAugspurger commented:
@datapythonista are you able to reproduce the CI failures like https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=12231 locally? If not, I can try taking a look.

datapythonista commented:
No, I'm taking a look, but very confused on what's going on, even the mypy error. The CI was green before merging master, which confuses me even more. If you can take a look too, that would be great, let me know if you find something.

WillAyd commented:
@datapythonista haven't had the chance to look too deep but interesting that `boxplot_frame` throws a mypy error while `hist_frame` the line above doesn't. What's different between those two is the Appender - can you see if removing that maybe silences the mypy error?

datapythonista commented:
thanks @WillAyd, I also saw that, and removing the Appender does fix the problem, but I think there is a root error, Appender is being used eveywhere... :\

WillAyd commented:
Yea not sure why it's appearing now. Probably just need to annotate Appender for proper inference.

I don't think it's critical for this PR so if that's the only hangup feel free to add `type: ignore` and we can address in a follow up

datapythonista commented:
I tried annotating `Appender` but didn't work. I saw a mypy issue where this error was raised for unrelated problems. One of the test failures (happens just with couple of builds) is related to boxplot too. I'll see if I can find the problem for that, as it's difficult to debug the mypy error. With some luck all the problems are caused by the same. Thanks for the help.

WillAyd commented:
No problem. As mentioned I think OK to mark as type: ignore for now so it doesn't hold this up. I've opened #26629 to address separately

datapythonista commented:
No idea what's going on here. The 3 errors seem unrelated. One of them is a failure on the qt backend of matplotlib because can't find the environment variable `DISPLAY` (I don't think this PR affects that variable, I guess the problem is something else causing that).

The other error is caused by a plot losing the tick information in a plot. None of them reproduce locally with the dependencies of the build (one is the mac build, may be someone with a mac can try to reproduce).

The mypy error seems to be nonsense too.

TomAugspurger commented:
I tried on a Mac, and wasn't able to reproduce the error (will try in an exact replica of that env later). I will note that the warning shows up under pytest under your branch, whereas it didn't before. I'm not sure why that would be.

Commit:
Merging from master

datapythonista committed:
Merge remote-tracking branch 'upstream/master' into plot

datapythonista committed:
Avoiding recursive import

datapythonista committed:
Add missing register/deregister to the public pandas-matplotlib API

datapythonista committed:
registering converters

datapythonista committed:
isort

datapythonista commented:
I fixed the errors, not sure why they were caused, but importing matplotlib to see if it was installed before registering the converters was causing the problems (including the mypy one which is weird).

CI is green, and it's generating the warning reported by @TomAugspurger like in master. Should be ready to be merged.

jreback commented:
merge master one more time, ping on green.

datapythonista committed:
Merging from master

jreback commented on pandas/plotting/_matplotlib/boxplot.py:
```diff
@@ -0,0 +1,339 @@
+
+    def _make_plot(self):
+        if self.subplots:
+            from pandas.core.series import Series
```
later on I suppose could move all imports to the top as this is now a delayed import generally

jreback commented:
@datapythonista after merge if you can create any needed followup issues, and issues for splitting tests and moving imports

datapythonista commented:
@jreback this is green.

Will open PRs for the immediate follow up things, and create issues for the rest once merged.

jreback commented:
thanks @datapythonista very nice!
